### PR TITLE
Speed up joins on indexed tables with string keys

### PR DESCRIPTION
**Pull request by ccaominh**
### Description

When joining on index tables with string keys, caching the computation of row id to row numbers improves performance on the `JoinAndLookupBenchmark.joinIndexTableStringKey`* benchmarks by about 10% if the column cache is enabled an by about 100% if the column cache is disabled. 

#### Before

```
Benchmark                            (columnCacheSizeBytes)   Score   Error  Units
joinIndexedTableStringKey                                 0  41.899 ± 0.688  ms/op
joinIndexedTableStringKey                             16384  22.707 ± 0.309  ms/op
joinIndexedTableStringKeyWithFilter                       0  41.879 ± 0.507  ms/op
joinIndexedTableStringKeyWithFilter                   16384  22.314 ± 0.114  ms/op
```

#### After

```
Benchmark                            (columnCacheSizeBytes)   Score   Error  Units
joinIndexedTableStringKey                                 0  20.527 ± 0.751  ms/op
joinIndexedTableStringKey                             16384  20.804 ± 0.206  ms/op
joinIndexedTableStringKeyWithFilter                       0  21.374 ± 0.299  ms/op
joinIndexedTableStringKeyWithFilter                   16384  19.723 ± 0.390  ms/op
```

(See https://github.com/apache/druid/pull/9267 for the `JoinAndLookupBenchmark` implementation.)

<hr>

This PR has:
- [x] been self-reviewed.
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.

### CODE REVIEW

**Comment by gianm** [comment]
How do these numbers compare to the standard lookup case?

File: processing/src/main/java/org/apache/druid/segment/join/table/IndexedTableJoinMatcher.java [hunk 1]
```diff
@@ -262,8 +288,9 @@ public ValueType defaultType()
         final IndexedInts row = selector.getRow();
 
         if (row.size() == 1) {
-          final String key = selector.lookupName(row.get(0));
-          return index.find(key).iterator();
+          int dimensionId = row.get(0);
+          //noinspection ConstantConditions (cache cannot return nulls since nulls are never stored in cache)
+          return dimensionCaches.get(selector).get(dimensionId).iterator();
```
Comment (gianm) [inline processing/src/main/java/org/apache/druid/segment/join/table/IndexedTableJoinMatcher.java]: This is only safe when the selector has a 'real' dictionary, i.e. when `selector.getValueCardinality()` does not return `DimensionSelector.CARDINALITY_UNKNOWN`. If it is unknown then the ids are not valid outside the context of a specific row. So, in that case you'll need to fall back to the slower code.

File: processing/src/main/java/org/apache/druid/segment/join/table/IndexedTableJoinMatcher.java [hunk 2]
```diff
@@ -240,13 +243,36 @@ private void advanceCurrentRow()
...
+                          Caffeine.newBuilder()
+                                  .maximumSize(CACHE_MAX_SIZE)
+                                  .build(dimensionId -> getRowNumbers(selector, dimensionId))
+                  );
+
+    }
+
+    private IntList getRowNumbers(DimensionSelector selector, int dimensionId)
+    {
+      final String key = selector.lookupName(dimensionId);
```
Comment (gianm) [inline processing/src/main/java/org/apache/druid/segment/join/table/IndexedTableJoinMatcher.java]: For each selector, `dimensionId`s are contiguous and go from `0` (inclusive) to `selector.getValueCardinality()` (exclusive). If the cardinality is less than `CACHE_MAX_SIZE` then you could use an `IntList[]` as a cache. It should be faster.

File: processing/src/main/java/org/apache/druid/segment/join/table/IndexedTableJoinMatcher.java [hunk 3]
```diff
@@ -240,13 +243,36 @@ private void advanceCurrentRow()
...
     {
       this.keyType = keyType;
       this.index = index;
+
+      this.dimensionCaches =
+          Caffeine.newBuilder()
+                  .maximumSize(MAX_NUM_CACHE)
+                  .build(
+                      selector ->
+                          Caffeine.newBuilder()
```
Comment (gianm) [inline processing/src/main/java/org/apache/druid/segment/join/table/IndexedTableJoinMatcher.java]: How does this compare to the LruEvalCache in SingleStringInputCachingExpressionColumnValueSelector? I think they're trying to solve the same problem, so whichever approach works better, both that class and this one should use the best approach.

**Comment by ccaominh** [comment]
> How do these numbers compare to the standard lookup case?

Lookups are about 5x faster:

```
Benchmark                               (columnCacheSizeBytes)  Score   Error  Units
lookupVirtualColumnStringKey                                 0  3.374 ± 0.065  ms/op
lookupVirtualColumnStringKey                             16384  3.364 ± 0.054  ms/op
lookupVirtualColumnStringKeyWithFilter                       0  4.553 ± 0.114  ms/op
lookupVirtualColumnStringKeyWithFilter                   16384  5.529 ± 0.305  ms/op
```

File: processing/src/main/java/org/apache/druid/segment/join/table/IndexedTableJoinMatcher.java (hunk 2 above)
Comment (gianm) [reply processing/src/main/java/org/apache/druid/segment/join/table/IndexedTableJoinMatcher.java]: (Assuming the dictionary is real, i.e., `selector.getValueCardinality() != DimensionSelector.CARDINALITY_UNKNOWN`)

File: processing/src/main/java/org/apache/druid/segment/join/table/IndexedTableJoinMatcher.java (hunk 3 above)
Comment (ccaominh) [reply processing/src/main/java/org/apache/druid/segment/join/table/IndexedTableJoinMatcher.java]: On `JoinAndLookupBenchmark.joinIndexedTableStringKey`, using the `LruEvalCache` approach is about 10% faster than using `Caffeine`.

**Comment by ccaominh** [comment]
Benchmark numbers with the improved cache implementation:

#### Original
```
Benchmark                               (columnCacheSizeBytes)   Score   Error  Units
joinIndexedTableStringKey                                    0  42.339 ± 0.323  ms/op
joinIndexedTableStringKey                                16384  22.634 ± 0.247  ms/op
joinIndexedTableStringKeyWithFilter                          0  38.170 ± 0.268  ms/op
joinIndexedTableStringKeyWithFilter                      16384  24.974 ± 0.462  ms/op
lookupVirtualColumnStringKey                                 0   4.040 ± 0.042  ms/op
lookupVirtualColumnStringKey                             16384   3.194 ± 0.011  ms/op
lookupVirtualColumnStringKeyWithFilter                       0   5.288 ± 0.027  ms/op
lookupVirtualColumnStringKeyWithFilter                   16384   5.175 ± 0.042  ms/op
```

#### Int2IntListLruCache
```
Benchmark                               (columnCacheSizeBytes)   Score   Error  Units
joinIndexedTableStringKey                                    0  14.557 ± 0.151  ms/op
joinIndexedTableStringKey                                16384  15.167 ± 0.075  ms/op
joinIndexedTableStringKeyWithFilter                          0  16.093 ± 0.517  ms/op
joinIndexedTableStringKeyWithFilter                      16384  16.434 ± 0.185  ms/op
```

#### Int2IntListLookupTable
```
Benchmark                               (columnCacheSizeBytes)   Score   Error  Units
joinIndexedTableStringKey                                    0  12.759 ± 0.125  ms/op
joinIndexedTableStringKey                                16384  12.439 ± 0.481  ms/op
joinIndexedTableStringKeyWithFilter                          0  14.139 ± 0.127  ms/op
joinIndexedTableStringKeyWithFilter                      16384  14.163 ± 0.176  ms/op
```

File: processing/src/main/java/org/apache/druid/segment/join/table/IndexedTableJoinMatcher.java [hunk 4]
```diff
@@ -263,8 +299,17 @@ public ValueType defaultType()
         final IndexedInts row = selector.getRow();
 
         if (row.size() == 1) {
-          final String key = selector.lookupName(row.get(0));
-          return index.find(key).iterator();
+          int dimensionId = row.get(0);
+          final IntList rowNumbers;
+          if (selector.getValueCardinality() == DimensionDictionarySelector.CARDINALITY_UNKNOWN) {
```
Comment (gianm) [inline processing/src/main/java/org/apache/druid/segment/join/table/IndexedTableJoinMatcher.java]: This won't change from row to row, so it's likely going to be better to check it outside hot code (i.e. return a different `Supplier<IntIterator>` based on the result of calling `selector.getValueCardinality()`).

File: processing/src/main/java/org/apache/druid/segment/join/table/IndexedTableJoinMatcher.java [hunk 5]
```diff
@@ -308,4 +353,112 @@ public ValueType defaultType()
...
+  static class Int2IntListLookupTable implements Int2IntListMap
+  {
+    private final IntList[] lookup;
+    private final IntFunction<IntList> loader;
+
+    Int2IntListLookupTable(int maxSize, IntFunction<IntList> loader)
+    {
+      this.loader = loader;
+      this.lookup = new IntList[maxSize];
+      Arrays.fill(lookup, null);
```
Comment (gianm) [inline processing/src/main/java/org/apache/druid/segment/join/table/IndexedTableJoinMatcher.java]: I thought Java guaranteed that object arrays will start with all `null`s.

**Pull request merged**